Windows service handling improvements

We need to think about how users are supposed to upgrade their sidecar to the
new version.

Right now a user cannot simply install the new sidecar (1.0) over an existing
older installation.
One problem is that the windows `service name` got changed, but not the 
`Display name`.
That's why the new sidecar is not able to register itself as a service,
nor is it able to uninstall the old sidecar's service.

There are probably more issues, we can add here...

> @mpfz0r Can we change the `Display name`?

I think that would improve things slightly.
However, I just noticed the same problem if you rename a collector:
```
time="2018-12-13T05:17:37-08:00" level=info msg="Adding process runner for: nxlog-windows-old" 
time="2018-12-13T05:17:37-08:00" level=info msg="[nxlog-windows-old] Configuration change detected, rewriting configuration file." 
time="2018-12-13T05:17:37-08:00" level=error msg="[nxlog-windows-old] Can't get status of service graylog-collector-nxlog-windows-old cause it doesn't exist: The specified service does not exist as an installed service." 
time="2018-12-13T05:17:37-08:00" level=error msg="[nxlog-windows-old] Failed to install service: The name is already in use as either a service name or a service display name." 
panic: runtime error: invalid memory address or nil pointer dereference
        panic: runtime error: invalid memory address or nil pointer dereference
[signal 0xc0000005 code=0x0 addr=0x10 pc=0x52c459]
```

--- a/daemon/daemon.go
+++ b/daemon/daemon.go
@@ -52,7 +52,7 @@ func NewConfig() *DaemonConfig {
 
 	dc := &DaemonConfig{
 		Name:        "graylog-sidecar",
-		DisplayName: "Graylog collector sidecar",
+		DisplayName: "Graylog Sidecar",
 		Description: "Wrapper service for Graylog controlled collector",
 		Dir:         rootDir,
 		Env:         []string{},
@@ -130,11 +130,18 @@ func (dc *DaemonConfig) SyncWithAssignments(configChecksums map[string]string, c
 			configChecksums[backend.Id] = ""
 		}
 	}
-
-	// add new backends to registry
+	assignedBackends := []*backends.Backend{}
 	for backendId := range assignments.Store.GetAll() {
 		backend := backends.Store.GetBackendById(backendId)
-		if backend != nil && dc.Runner[backend.Id] == nil {
+		if backend != nil {
+			assignedBackends = append(assignedBackends, backend)
+		}
+	}
+	CleanOldServices(assignedBackends)
+
+	// add new backends to registry
+	for _, backend := range assignedBackends {
+		if dc.Runner[backend.Id] == nil {
 			log.Info("Adding process runner for: " + backend.Name)
 			dc.AddRunner(*backend, context)
 		}

--- a/daemon/exec_runner.go
+++ b/daemon/exec_runner.go
@@ -193,7 +193,13 @@ func (r *ExecRunner) start() error {
 	}
 
 	// setup process environment
-	quotedArgs, err := shlex.Split(r.args)
+	var err error
+	var quotedArgs []string
+	if runtime.GOOS == "windows" {
+		quotedArgs = common.CommandLineToArgv(r.args)
+	} else {
+		quotedArgs, err = shlex.Split(r.args)
+	}
 	if err != nil {
 		return err
 	}

--- a/daemon/svc_helper.go
+++ b/daemon/svc_helper.go
@@ -1,0 +1,11 @@
+// +build !windows
+
+package daemon
+
+import (
+	"github.com/Graylog2/collector-sidecar/backends"
+)
+
+// Dummy function. Only used on Windows
+func CleanOldServices(assignedBackends []*backends.Backend) {
+}

--- a/daemon/svc_helper_windows.go
+++ b/daemon/svc_helper_windows.go
@@ -1,0 +1,113 @@
+package daemon
+
+import (
+	"fmt"
+	"github.com/Graylog2/collector-sidecar/backends"
+	"golang.org/x/sys/windows/svc"
+	"golang.org/x/sys/windows/svc/eventlog"
+	"golang.org/x/sys/windows/svc/mgr"
+	"strings"
+	"time"
+)
+
+func CleanOldServices(assignedBackends []*backends.Backend) {
+	registeredServices, err := getRegisteredServices()
+	if err != nil {
+		log.Warn("Failed to get registered services. Skipping cleanup. ", err)
+		return
+	}
+	for _, service := range registeredServices {
+		if strings.Contains(service, ServiceNamePrefix) {
+			log.Debugf("Found graylog service %s", service)
+			if !serviceIsAssigned(assignedBackends, service) {
+				log.Infof("Removing stale graylog service %s", service)
+				uninstallService(service)
+			}
+		}
+	}
+}
+
+func getRegisteredServices() ([]string, error) {
+	m, err := mgr.Connect()
+	if err != nil {
+		return nil, err
+	}
+	defer m.Disconnect()
+	registeredServices, err := m.ListServices()
+	if err != nil {
+		return nil, err
+	}
+	return registeredServices, nil
+}
+
+func serviceIsAssigned(assignedBackends []*backends.Backend, serviceName string) bool {
+	for _, backend := range assignedBackends {
+		if backend.ServiceType == "svc" && ServiceNamePrefix+backend.Name == serviceName {
+			return true
+		}
+	}
+	return false
+}
+
+func uninstallService(name string) {
+	log.Infof("Uninstalling service %s", name)
+	stopService(name)
+
+	m, err := mgr.Connect()
+	if err != nil {
+		log.Errorf("Failed to connect to service manager: %v", err)
+		return
+	}
+	defer m.Disconnect()
+
+	s, err := m.OpenService(name)
+	// service exist so we try to uninstall it
+	if err != nil {
+		log.Debugf("Service %s doesn't exist, no uninstall action needed", name)
+		return
+	}
+
+	defer s.Close()
+	err = s.Delete()
+	if err != nil {
+		log.Errorf("Can't delete service %s: %v", name, err)
+	}
+
+	err = eventlog.Remove(s.Name)
+	if err != nil {
+		log.Errorf("RemoveEventLogSource() failed: %s", err)
+	}
+	return
+}
+
+func stopService(serviceName string) error {
+	m, err := mgr.Connect()
+	if err != nil {
+		return fmt.Errorf("Failed to connect to service manager: %v", err)
+	}
+	defer m.Disconnect()
+
+	ws, err := m.OpenService(serviceName)
+	if err != nil {
+		return fmt.Errorf("Could not access service %s: %v", serviceName, err)
+	}
+	defer ws.Close()
+
+	status, err := ws.Control(svc.Stop)
+	if err != nil {
+		return fmt.Errorf("Could not send stop control: %v", err)
+	}
+
+	timeout := time.Now().Add(10 * time.Second)
+	for status.State != svc.Stopped {
+		if timeout.Before(time.Now()) {
+			return fmt.Errorf("Timeout waiting for service to go to stopped state: %v", err)
+		}
+		time.Sleep(300 * time.Millisecond)
+		status, err = ws.Query()
+		if err != nil {
+			return fmt.Errorf("Could not retrieve service status: %v", err)
+		}
+	}
+	return nil
+}

--- a/daemon/svc_runner_windows.go
+++ b/daemon/svc_runner_windows.go
@@ -81,9 +81,7 @@ func (r *SvcRunner) Running() bool {
 	defer m.Disconnect()
 
 	s, err := m.OpenService(r.serviceName)
-	// service exist so we only update the properties
 	if err != nil {
-		r.backend.SetStatusLogErrorf("Can't get status of service %s cause it doesn't exist: %v", r.serviceName, err)
 		return false
 	}
 	defer s.Close()
@@ -91,6 +89,7 @@ func (r *SvcRunner) Running() bool {
 	status, err := s.Query()
 	if err != nil {
 		r.backend.SetStatusLogErrorf("Can't query status of service %s: %v", r.serviceName, err)
+		return false
 	}
 
 	return status.State == svc.Running
@@ -127,8 +126,7 @@ func (r *SvcRunner) ValidateBeforeStart() error {
 		return err
 	}
 
-	execPath, err := exec.LookPath(r.exec)
-	if err != nil {
+	if _, err := exec.LookPath(r.exec); err != nil {
 		return r.backend.SetStatusLogErrorf("Failed to find collector executable %s", r.exec)
 	}
 
@@ -161,10 +159,20 @@ func (r *SvcRunner) ValidateBeforeStart() error {
 		// service needs to be created
 	} else {
 		s, err = m.CreateService(r.serviceName,
-			execPath,
+			r.exec,
 			serviceConfig)
 		if err != nil {
 			return r.backend.SetStatusLogErrorf("Failed to install service: %v", err)
+		}
+		// It seems impossible to create a service with properly quoted arguments :-(
+		// Updating the BinaryPathName afterwards does the trick
+		currentConfig, err := s.Config()
+		if err == nil {
+			currentConfig.BinaryPathName = serviceConfig.BinaryPathName
+		}
+		err = s.UpdateConfig(currentConfig)
+		if err != nil {
+			r.backend.SetStatusLogErrorf("Failed to update the created service: %v", err)
 		}
 		defer s.Close()
 		err = eventlog.InstallAsEventCreate(r.serviceName, eventlog.Error|eventlog.Warning|eventlog.Info)

--- a/daemon/svc_runner_windows.go
+++ b/daemon/svc_runner_windows.go
@@ -28,6 +28,8 @@ import (
 	"github.com/Graylog2/collector-sidecar/context"
 )
 
+const ServiceNamePrefix = "graylog-collector-"
+
 type SvcRunner struct {
 	RunnerCommon
 	exec         string
@@ -54,7 +56,7 @@ func NewSvcRunner(backend backends.Backend, context *context.Ctx) Runner {
 		exec:        backend.ExecutablePath,
 		args:        backend.ExecuteParameters,
 		signals:     make(chan string),
-		serviceName: "graylog-collector-" + backend.Name,
+		serviceName: ServiceNamePrefix + backend.Name,
 	}
 
 	// set default state
@@ -113,6 +115,7 @@ func (r *SvcRunner) GetBackend() *backends.Backend {
 func (r *SvcRunner) SetBackend(b backends.Backend) {
 	r.backend = b
 	r.name = b.Name
+	r.serviceName = ServiceNamePrefix + b.Name
 	r.exec = b.ExecutablePath
 	r.args = b.ExecuteParameters
 }
@@ -137,14 +140,14 @@ func (r *SvcRunner) ValidateBeforeStart() error {
 
 	serviceConfig := mgr.Config{
 		DisplayName:    "Graylog collector sidecar - " + r.name + " backend",
-		Description:    "Wrapper service for the NXLog backend",
+		Description:    "Wrapper service for the " + r.name + " backend",
 		BinaryPathName: "\"" + r.exec + "\" " + r.args}
 
 	s, err := m.OpenService(r.serviceName)
 	// service exist so we only update the properties
 	if err == nil {
 		defer s.Close()
-		log.Debugf("[%s] service %s already exists, updating properties", r.name)
+		log.Debugf("Service %s already exists, updating properties", r.name)
 		currentConfig, err := s.Config()
 		if err == nil {
 			currentConfig.DisplayName = serviceConfig.DisplayName
@@ -161,13 +164,13 @@ func (r *SvcRunner) ValidateBeforeStart() error {
 			execPath,
 			serviceConfig)
 		if err != nil {
-			r.backend.SetStatusLogErrorf("Failed to install service: %v", err)
+			return r.backend.SetStatusLogErrorf("Failed to install service: %v", err)
 		}
 		defer s.Close()
 		err = eventlog.InstallAsEventCreate(r.serviceName, eventlog.Error|eventlog.Warning|eventlog.Info)
 		if err != nil {
 			s.Delete()
-			r.backend.SetStatusLogErrorf("SetupEventLogSource() failed: %v", err)
+			return r.backend.SetStatusLogErrorf("SetupEventLogSource() failed: %v", err)
 		}
 	}
 
@@ -213,7 +216,7 @@ func (r *SvcRunner) start() error {
 
 	ws, err := m.OpenService(r.serviceName)
 	if err != nil {
-		return r.backend.SetStatusLogErrorf("Could not access service: %v", err)
+		return r.backend.SetStatusLogErrorf("Could not access service %s: %v", r.serviceName, err)
 	}
 	defer ws.Close()
 
@@ -239,35 +242,10 @@ func (r *SvcRunner) stop() error {
 	// deactivate supervisor
 	r.setSupervised(false)
 
-	m, err := mgr.Connect()
+	err := stopService(r.serviceName)
 	if err != nil {
-		return r.backend.SetStatusLogErrorf("Failed to connect to service manager: %v", err)
+		return r.backend.SetStatusLogErrorf("%s", err)
 	}
-	defer m.Disconnect()
-
-	ws, err := m.OpenService(r.serviceName)
-	if err != nil {
-		return r.backend.SetStatusLogErrorf("Could not access service: %v", err)
-	}
-	defer ws.Close()
-
-	status, err := ws.Control(svc.Stop)
-	if err != nil {
-		return r.backend.SetStatusLogErrorf("Could not send stop control: %v", err)
-	}
-
-	timeout := time.Now().Add(10 * time.Second)
-	for status.State != svc.Stopped {
-		if timeout.Before(time.Now()) {
-			return r.backend.SetStatusLogErrorf("Timeout waiting for service to go to stopped state: %v", err)
-		}
-		time.Sleep(300 * time.Millisecond)
-		status, err = ws.Query()
-		if err != nil {
-			return r.backend.SetStatusLogErrorf("Could not retrieve service status: %v", err)
-		}
-	}
-
 	r.backend.SetStatus(backends.StatusStopped, "Stopped", "")
 
 	return nil

--- a/glide.lock
+++ b/glide.lock
@@ -32,7 +32,7 @@ imports:
   subpackages:
   - yaml
 - name: golang.org/x/sys
-  version: 833a04a10549a95dc34458c195cbad61bbb6cb4d
+  version: 98c5dad5d1a0e8a73845ecc8897d0bd56586511d
   subpackages:
   - unix
   - windows

--- a/services/control_handler_windows.go
+++ b/services/control_handler_windows.go
@@ -16,10 +16,9 @@
 package services
 
 import (
+	"github.com/Graylog2/collector-sidecar/backends"
+	"github.com/Graylog2/collector-sidecar/daemon"
 	"github.com/kardianos/service"
-
-	"golang.org/x/sys/windows/svc/eventlog"
-	"golang.org/x/sys/windows/svc/mgr"
 )
 
 func ControlHandler(action string) {
@@ -37,35 +36,6 @@ func ControlHandler(action string) {
 }
 
 func cleanupInstalledServices() {
-	// backends are not initialized yet, let's use names directly
-	uninstallService("graylog-collector-nxlog")
-}
-
-func uninstallService(name string) {
-	log.Infof("Uninstalling service %s", name)
-	m, err := mgr.Connect()
-	if err != nil {
-		log.Errorf("Failed to connect to service manager: %v", err)
-	}
-	defer m.Disconnect()
-
-	s, err := m.OpenService(name)
-	// service exist so we try to uninstall it
-	if err != nil {
-		log.Debugf("Service %s doesn't exist, no uninstall action needed", name)
-		return
-	}
-
-	defer s.Close()
-	err = s.Delete()
-	if err != nil {
-		log.Errorf("Can't delete service %s: %v", name, err)
-	}
-
-	err = eventlog.Remove(s.Name)
-	if err != nil {
-		log.Errorf("RemoveEventLogSource() failed: %s", err)
-	}
-
-	return
+	// Cleans all services starting with "graylog-collector-"
+	daemon.CleanOldServices([]*backends.Backend{})
 }

